corehq/apps/app_manager/suite_xml/post_process/remote_requests.py

```diff
@@ -231,6 +234,25 @@ def _remote_request_query_datums(self):
             )
         return datums
 
+    # Enables app builders to use a special {} bracket syntax (mimicing python) to insert
+    # variables for interpretation by formplayer.
+    @staticmethod
+    def _convert_curly_braces_to_concat(xpath_query):
+
+        if '{' in xpath_query:
+            # If string uses double quotes remove them because we are going to add a concat.
```

Comment:
A user could also use outer single quotes an inner double quotes, right?

Reply:
There are also cases where a user may use either of the `{}` characters outside independently. I am starting to wonder if we aren't being naive about this: 

`if(contains(code, '{'), true, if(contains(code, '}'), true, false))`

Reply:
I think parsing this way is a bad idea and is bound to cause pain in the future. This job really requires a proper XPath parser, not simple find/replace.

Reply:
I think I need more context. Are we extending the XPath grammar with a new syntactical construct, or are we adding a templating system on top of XPath? Vellum does something like the former for [hashtag expressions](
https://github.com/dimagi/js-xpath#usage) (`#case/property`). In the latter case, think Jinja2 on top of HTML (or any other text document). If we're adding templating on top of XPath then it seems like we may want a way to denote that an expression should be interpreted first as a template before evaluating it as XPath. As you point out, it is possible that template primitives could exist in string literals, so they either need to be escaped or there needs to be some indication that the template system should not be applied to them.

Would it make sense to pick an existing template language such as the one used by Python's `string.format()` or f-strings rather than rolling our own if we go the templating route?

Reply:
> or are we adding a templating system on top of XPath

Something like this yes.

We could use an existing template language but we'd need to be able to get the raw parsed AST so that we can replace variable names with instance expressions and re-structure the expression (see https://github.com/dimagi/commcare-hq/pull/31204#issuecomment-1063845421).

I've looked into Jinja2 but it won't parse xpath expressions like `instance('casedb')/casedb/case[@case_type='patient']`, same for `mako`. The Django template Lexer could work as long as we're happy with the double curly braces syntax: `{{ }}`:

```python
from django.template.base import Lexer, TokenType
from eulxml.xpath import parse as parse_xpath
xpath_template = "name = '{{$name}}' and status = '{{instance('casedb')/casedb/case[@case_id='123']/status}}'"
for token in Lexer(xpath_template).tokenize():
    if token.token_type == TokenType.VAR:
        print('xpath', parse_xpath(token.contents))
    else:
        print(token.token_type, f'"{token.contents}"')

TokenType.TEXT "name = '"
xpath <VariableReference $name>
TokenType.TEXT "' and status = '"
xpath <BinaryExpression instance('casedb') / casedb/case[@case_id='123']/status>
TokenType.TEXT "'"
```

Reply:
What about collisions between the template syntax and unintentionally embedded template directives in XPath expressions. Is this why Jinja2 and Mako don't work? Repurposing your example above, how will this be handled?

```
if(contains(code, '{{'), true, if(contains(code, '}}'), true, false))
```
I plugged it into the Django template processor above and got
```
TokenType.TEXT "if(contains(code, '"
xpath ), true, if(contains(code, 
TokenType.TEXT "'), true, false))"
```

This is a general issue for any templating solution, not just Django templates (just substitute relevant delimiters).

Extending the XPath parser with  a new syntactical construct may be a better solution because it avoids this issue.

Reply:
> Extending the XPath parser with a new syntactical construct may be a better solution because it avoids this issue.

I don't think it is possible to extend the XPath parser without completely replacing it or doing some funny monkey patching. A separate parser is an option and something we already have a prototype of: https://github.com/dimagi/commcare-hq/pull/31164

For the example above I think we could handle that since the 'var' token parsed as a string literal so we could ignore it. Here's a trickier edge case (though quite contrived):

```
prop = '{{ b = "' and other = 'something" }}'

TokenType.TEXT "prop = '"
xpath <BinaryExpression b = "' and other = 'something">
TokenType.TEXT "'"
```